File: Allegro5 Test/Ghost.cpp
#include "Ghost.h"


Ghost::Ghost(void){}

void Ghost::Init(ALLEGRO_BITMAP *image, int spawnX, int spawnY, int patrol[2]){
	GameObject::Init(spawnX, spawnY, 3, 3, 1, 0, 8, 32);
	SetID(ENEMY);
	SetAlive(true);
	SetCollideType(WHITE);

	lives = 1;
	frameCount = 0;
	frameDelay = 5;
	frameHeight = 64;
	frameWidth = 64;
	animationColumns = 8;
	animationRow = 0;
	curFrame = 0;
	this->patrol[1] = patrol[1];
	this->patrol[0] = patrol[0];

	if(image != 0){
		Ghost::image = image;
	}

}

void Ghost::Update(){
	GameObject::Update();
	if(x <= patrol[0] || x >= patrol[1]){
		dirX = -dirX;
	}
	if(dirX == -1){
		MoveLeft();
	}
	else if(dirX == 1){
		MoveRight();
	}
	frameCount++;
}

void Ghost::Render(){
	GameObject::Render();

	int fx = (curFrame % animationColumns) * frameWidth;
	int fy = animationRow * frameHeight;

	al_draw_bitmap_region(image, fx, fy, frameWidth, frameHeight, x-frameWidth/2, y-frameHeight/2, 0);
}

void Ghost::MoveLeft(){
	if(frameCount>=frameDelay){
		frameCount = 0;
		if(animationRow!=5){
			animationRow = 5;
			curFrame = 7;
		}
		else if(curFrame == 4){
			curFrame = 7;
		}
		else --curFrame;
	}
}

void Ghost::MoveRight(){
	if(frameCount>=frameDelay){
		frameCount = 0;
		if(animationRow!=1){
			animationRow = 1;
			curFrame = 0;
		}
		else if(curFrame == 3)
			curFrame = 0;
		curFrame++;
	}
}

void Ghost::Attack(){

}

void Ghost::Die(){

}

File: js/ghost.js
// Ghost class - enemy that patrols between points

class Ghost extends GameObject {
    constructor() {
        super();
        this.setId(ID.ENEMY);
        this.setCollidable(true);
        
        this.lives = 1;
        this.animationRow = 0;
        this.patrol = [0, 0];
        this.patrolDirection = 1;
        this.moveSpeed = 1;
    }
    
    init(texture, spawnX, spawnY, patrolPoints) {
        super.init(spawnX, spawnY, 0, 0, 1, 0, 16, 16);
        
        this.patrol[0] = patrolPoints[0];
        this.patrol[1] = patrolPoints[1];
        
        if (texture) {
            this.sprite = new PIXI.Sprite(texture);
            this.sprite.width = 32;
            this.sprite.height = 32;
            this.sprite.anchor.set(0.5, 0.5);
            this.sprite.tint = 0x888888; // Ghost-like appearance
        }
        
        this.frameWidth = 32;
        this.frameHeight = 32;
    }
    
    update() {
        // Patrol between two points
        if (this.patrolDirection > 0) {
            this.x += this.moveSpeed;
            if (this.x >= this.patrol[1]) {
                this.patrolDirection = -1;
            }
        } else {
            this.x -= this.moveSpeed;
            if (this.x <= this.patrol[0]) {
                this.patrolDirection = 1;
            }
        }
        
        this.velX = this.moveSpeed * this.patrolDirection;
        
        // Update animation
        this.frameCount++;
        if (this.frameCount >= 10) { // Frame delay
            this.frameCount = 0;
            this.curFrame++;
            if (this.curFrame >= 4) { // Assuming 4 frames of animation
                this.curFrame = 0;
            }
        }
    }
    
    render() {
        super.render();
    }
    
    moveLeft() {
        this.patrolDirection = -1;
    }
    
    moveRight() {
        this.patrolDirection = 1;
    }
    
    attack() {
        // Ghost attack behavior
    }
    
    die() {
        this.setAlive(false);
    }
}